### ISSUE TITLE
[BUG] Binary release with CUDA failing in Google Colab environment

### ISSUE THREAD

**Issue by atolkachiov** [arrayfire/arrayfire#3142]
I am setting up google colab playground to run some experiments. 
Here is the notebook with helloworld example which fails when running with GPU kernel, while works fine on CPU: 
https://colab.research.google.com/drive/1xUVsyKJgD7pvPDRnhliYzzqjcctpK7E7?usp=sharing

```
ArrayFire Exception (Internal error:998):
In function void print(const char*, af_array, int, std::ostream&, bool) [with T = float; af_array = void*; std::ostream = std::basic_ostream<char>]
In file src/api/c/print.cpp:99

```

Description
===========
* Did you build ArrayFire yourself or did you use the official installers
Using binary installer: https://arrayfire.s3.amazonaws.com/3.8.0/ArrayFire-v3.8.0_Linux_x86_64.sh

* Which backend is experiencing this issue? (CPU, CUDA, OpenCL)
CUDA throws exception
CPU works as expected 

* Do you have a workaround?
No

* Can the bug be reproduced reliably on your system?
Yes

* Run your executable with AF_TRACE=all and AF_PRINT_ERRORS=1 environment
  variables set.
The relevant error is likely 
`Unable to open /root/.arrayfire/KER11230367656483596278_CU_60_AF_38.bin for Tesla P100-PCIE-16GB`
You can see full trace in the notebook. 

* Screenshot or terminal output of the results
N/A, check colab link. 

Reproducible Code and/or Steps
------------------------------
Easily reproducible, try running my shared colab notebook. 
To ensure you are running on GPU, select Change Runtime -> Change runtime type -> GPU. 

System Information
------------------
```
ArrayFire v3.8.0 (CUDA, 64-bit Linux, build d99887a)
Platform: CUDA Runtime 11.0, Driver: 460.32.03
[0] Tesla P100-PCIE-16GB, 16281 MB, CUDA Compute 6.0
```

Run one of the following commands based on your OS
```
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 18.04.5 LTS
Release:	18.04
Codename:	bionic
Failed to initialize NVML: Driver/library version mismatch
rocm-smi not found.
Number of platforms                               1
  Platform Name                                   NVIDIA CUDA
  Platform Vendor                                 NVIDIA Corporation
  Platform Version                                OpenCL 3.0 CUDA 11.3.101
  Platform Profile                                FULL_PROFILE
  Platform Extensions                             cl_khr_global_int32_base_atomics cl_khr_global_int32_extended_atomics cl_khr_local_int32_base_atomics cl_khr_local_int32_extended_atomics cl_khr_fp64 cl_khr_3d_image_writes cl_khr_byte_addressable_store cl_khr_icd cl_nv_compiler_options cl_nv_device_attribute_query cl_nv_pragma_unroll cl_nv_copy_opts cl_nv_create_buffer cl_khr_int64_base_atomics cl_khr_int64_extended_atomics cl_nv_kernel_attribute cl_khr_device_uuid
  Platform Host timer resolution                  0ns
  Platform Extensions function suffix             NV

  Platform Name                                   NVIDIA CUDA
Number of devices                                 1
  Device Name                                     Tesla P100-PCIE-16GB
  Device Vendor                                   NVIDIA Corporation
  Device Vendor ID                                0x10de
  Device Version                                  OpenCL 3.0 CUDA
  Driver Version                                  465.27
  Device OpenCL C Version                         OpenCL C 1.2 
  Device Type                                     GPU
  Device Topology (NV)                            PCI-E, 00:00.4
  Device Profile                                  FULL_PROFILE
  Device Available                                Yes
  Compiler Available                              Yes
  Linker Available                                Yes
  Max compute units                               56
  Max clock frequency                             1328MHz
  Compute Capability (NV)                         6.0
  Device Partition                                (core)
    Max number of sub-devices                     1
    Supported partition types                     None
  Max work item dimensions                        3
  Max work item sizes                             1024x1024x64
  Max work group size                             1024
  Preferred work group size multiple              32
  Warp size (NV)                                  32
  Max sub-groups per work group                   0
  Preferred / native vector sizes                 
    char                                                 1 / 1       
    short                                                1 / 1       
    int                                                  1 / 1       
    long                                                 1 / 1       
    half                                                 0 / 0        (n/a)
    float                                                1 / 1       
    double                                               1 / 1        (cl_khr_fp64)
  Half-precision Floating-point support           (n/a)
  Single-precision Floating-point support         (core)
    Denormals                                     Yes
    Infinity and NANs                             Yes
    Round to nearest                              Yes
    Round to zero                                 Yes
    Round to infinity                             Yes
    IEEE754-2008 fused multiply-add               Yes
    Support is emulated in software               No
    Correctly-rounded divide and sqrt operations  Yes
  Double-precision Floating-point support         (cl_khr_fp64)
    Denormals                                     Yes
    Infinity and NANs                             Yes
    Round to nearest                              Yes
    Round to zero                                 Yes
    Round to infinity                             Yes
    IEEE754-2008 fused multiply-add               Yes
    Support is emulated in software               No
  Address bits                                    64, Little-Endian
  Global memory size                              17071734784 (15.9GiB)
  Error Correction support                        Yes
  Max memory allocation                           4267933696 (3.975GiB)
  Unified memory for Host and Device              No
  Integrated memory (NV)                          No
  Shared Virtual Memory (SVM) capabilities        (core)
    Coarse-grained buffer sharing                 Yes
    Fine-grained buffer sharing                   No
    Fine-grained system sharing                   No
    Atomics                                       No
  Minimum alignment for any data type             128 bytes
  Alignment of base address                       4096 bits (512 bytes)
  Preferred alignment for atomics                 
    SVM                                           0 bytes
    Global                                        0 bytes
    Local                                         0 bytes
  Max size for global variable                    0
  Preferred total size of global vars             0
  Global Memory cache type                        Read/Write
  Global Memory cache size                        1376256 (1.312MiB)
  Global Memory cache line size                   128 bytes
  Image support                                   Yes
    Max number of samplers per kernel             32
    Max size for 1D images from buffer            268435456 pixels
    Max 1D or 2D image array size                 2048 images
    Max 2D image size                             16384x32768 pixels
    Max 3D image size                             16384x16384x16384 pixels
    Max number of read image args                 256
    Max number of write image args                16
    Max number of read/write image args           0
  Max number of pipe args                         0
  Max active pipe reservations                    0
  Max pipe packet size                            0
  Local memory type                               Local
  Local memory size                               49152 (48KiB)
  Registers per block (NV)                        65536
  Max number of constant args                     9
  Max constant buffer size                        65536 (64KiB)
  Max size of kernel argument                     4352 (4.25KiB)
  Queue properties (on host)                      
    Out-of-order execution                        Yes
    Profiling                                     Yes
  Queue properties (on device)                    
    Out-of-order execution                        No
    Profiling                                     No
    Preferred size                                0
    Max size                                      0
  Max queues on device                            0
  Max events on device                            0
  Prefer user sync for interop                    No
  Profiling timer resolution                      1000ns
  Execution capabilities                          
    Run OpenCL kernels                            Yes
    Run native kernels                            No
    Sub-group independent forward progress        No
    Kernel execution timeout (NV)                 No
  Concurrent copy and kernel execution (NV)       Yes
    Number of async copy engines                  2
    IL version                                    
  printf() buffer size                            1048576 (1024KiB)
```

**Comment by umar456** [comment]
Thanks for the link to your notebook. It really helps with the debugging of this issue.

Looks like the nvidia runtime is looking for a different nvrtc-buildtin library than provided by the installer. This is strange because we have tested this sceanario in our CI and it has been working correctly. I will need to investigate why this is happening. in the meantime you can work around this bug by renaming the
`/opt/arrayfire/lib64/libnvrtc-builtins.so` to `/opt/arrayfire/lib64/libnvrtc-builtins.so.11.1`

**Comment by atolkachiov** [comment]
@umar456 thanks for checking this out! It looks like CUDA 11.* is not supported in the binary installer https://arrayfire.s3.amazonaws.com/3.8.0/ArrayFire-v3.8.0_Linux_x86_64.sh 
You can see the error in my colab notebook: 
```
[platform][1623062663][000247] [ ../src/backend/cuda/device_manager.cpp:466 ] CUDA Driver supports up to CUDA 11.3 ArrayFire CUDA Runtime 11.0
[platform][1623062663][000247] [ ../src/backend/cuda/device_manager.cpp:451 ] CUDA driver version(11.3) not part of the CudaToDriverVersion array. Please create an issue or a pull request on the ArrayFire repository to update the CudaToDriverVersion variable with this version of the CUDA runtime.
```

Today I tried building from source in another colab and it worked, though it took hours to build everything so this is not practical to use in colab. 

As I see you have different CUDA 10 and 11 distributions, maybe you can just build one with CUDA 11 support for Linux? 
http://arrayfire.s3.amazonaws.com/index.html#!/3.8.0%2F

Having nightly builds would also be awesome: http://arrayfire.s3.amazonaws.com/index.html#!/nightly%2F

**Comment by umar456** [comment]
That message is not an error. It is targeted towards ArrayFire developers to update the minimum driver requirements for the latest CUDA toolkits. CUDA maintains a forward compatibility and therefore older versions of the toolkit would work and do not need intervention. We use the toolkit versions to display useful error message if the CUDA driver doesn't support a certain version of the CUDA toolkit. Since we do not know the minimum driver versions of the future versions of CUDA we allow the program to execute without those checks if the Toolkit version is not found in the database. In those cases we add the message to only the AF_TRACE logs. If you remove the AF_TRACE environment variable, you will not see that message. The toolkit version has also been updated in the latest master and should be part of the next release(see https://github.com/arrayfire/arrayfire/pull/3125).

**Comment by atolkachiov** [comment]
> That message is not an error. It is targeted towards ArrayFire developers to update the minimum driver requirements for the latest CUDA toolkits. CUDA maintains a forward compatibility and therefore older versions of the toolkit would work and do not need intervention. We use the toolkit versions to display useful error message if the CUDA driver doesn't support a certain version of the CUDA toolkit. Since we do not know the minimum driver versions of the future versions of CUDA we allow the program to execute without those checks if the Toolkit version is not found in the database. In those cases we add the message to only the AF_TRACE logs. If you remove the AF_TRACE environment variable, you will not see that message. The toolkit version has also been updated in the latest master and should be part of the next release(see #3125).

It makes sense, thanks! Adding this line fixed the issue in my colab: 
```
cp /opt/arrayfire/lib64/libnvrtc-builtins.so /opt/arrayfire/lib64/libnvrtc-builtins.so.11.1
```

**Comment by 9prady9** [comment]
Closing this as duplicate of issue fixed via #3105.
